tests/rust/wasm32-wasip3/src/bin/filesystem-symbolic-links.rs

```diff
@@ -0,0 +1,115 @@
+    generate_all
+});
+
+use wasi::filesystem::types::Descriptor;
+use wasi::filesystem::types::{DescriptorFlags, ErrorCode, OpenFlags, PathFlags};
+
+async fn test_symbolic_links(dir: &Descriptor) {
+    let ln_s = |from: &str, to: &str| -> _ { dir.symlink_at(from.to_string(), to.to_string()) };
+    let readlink = |path: &str| dir.readlink_at(path.to_string());
+    let stat_with_flags = |flags: PathFlags, path: &str| dir.stat_at(flags, path.to_string());
+    let stat_follow = |path: &str| stat_with_flags(PathFlags::SYMLINK_FOLLOW, path);
+    let open_r_follow = |path: &str| -> _ {
+        dir.open_at(
+            PathFlags::SYMLINK_FOLLOW,
+            path.to_string(),
+            OpenFlags::empty(),
+            DescriptorFlags::READ,
+        )
+    };
+    let open_r = |path: &str| -> _ {
+        dir.open_at(
+            PathFlags::empty(),
+            path.to_string(),
+            OpenFlags::empty(),
+            DescriptorFlags::READ,
+        )
+    };
+    let rm = |path: &str| dir.unlink_file_at(path.to_string());
+
+    // readlink-at: async func(path: string) -> result<string, error-code>;
+    assert_eq!(readlink("").await, Err(ErrorCode::NoEntry));
+    assert_eq!(readlink(".").await, Err(ErrorCode::Invalid));
+    assert_eq!(readlink("a.txt").await, Err(ErrorCode::Invalid));
+    assert_eq!(readlink("z.txt").await, Err(ErrorCode::NoEntry));
+    assert_eq!(readlink("./../").await, Err(ErrorCode::NotPermitted));
+    assert_eq!(readlink("..").await, Err(ErrorCode::NotPermitted));
+    assert_eq!(readlink("/").await, Err(ErrorCode::NotPermitted));
+    assert_eq!(readlink("parent").await, Ok("..".to_string()));
+
+    let afd = open_r("a.txt").await.unwrap();
+    assert_eq!(
+        afd.readlink_at(".".to_string()).await,
+        Err(ErrorCode::NotDirectory)
+    );
+
+    // https://github.com/WebAssembly/wasi-filesystem/issues/186
+    assert_eq!(
+        open_r("parent")
+            .await
+            .expect_err("open symlink with NOFOLLOW"),
+        ErrorCode::Loop
+    );
+
+    ln_s("parent", "parent-link").await.unwrap();
+    assert_eq!(open_r("parent-link").await.unwrap_err(), ErrorCode::Loop);
+    assert_eq!(
+        open_r_follow("parent-link").await.unwrap_err(),
+        ErrorCode::NotPermitted
+    );
+    assert_eq!(
+        stat_follow("parent-link").await.unwrap_err(),
+        ErrorCode::NotPermitted
+    );
+    assert_eq!(open_r("parent-link").await.unwrap_err(), ErrorCode::Loop);
+    assert_eq!(
+        ln_s("a.txt", "parent-link/a.txt").await,
+        Err(ErrorCode::NotPermitted)
+    );
+    rm("parent-link").await.unwrap();
+
+    ln_s("self", "self").await.unwrap();
+    assert_eq!(open_r_follow("self").await.unwrap_err(), ErrorCode::Loop);
+    assert_eq!(stat_follow("self").await.unwrap_err(), ErrorCode::Loop);
+    rm("self").await.unwrap();
+
+    assert_eq!(ln_s("whatever", "").await, Err(ErrorCode::NoEntry));
+    assert_eq!(ln_s("", "whatever").await, Err(ErrorCode::NoEntry));
```

Comment:
Wasmtime MacOS failure: 
Test filesystem-symbolic-links failed
  [exit_code] 0 == 134
STDOUT:

STDERR:

thread '<unnamed>' (1) panicked at src/bin/filesystem-symbolic-links.rs:92:5:
assertion `left == right` failed
  left: Ok(())
 right: Err(ErrorCode { code: 19, name: "no-entry", message: "No such file or directory, similar to `ENOENT` in POSIX." })
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Error: failed to run main module `tests/rust/testsuite/wasm32-wasip3/filesystem-symbolic-links.wasm`